### ISSUE TITLE
Add escape function

### ISSUE THREAD

**Issue by florianpircher** [fancy-regex/fancy-regex#60]
The `regex` crates provides an [`escape(text: &str) -> String` function](https://docs.rs/regex/1.4.1/regex/fn.escape.html) which “escapes all regular expression meta characters in text”. This is especially useful when the regex pattern contains a dynamic value such as user-input or a network-response.

For example, a search that includes the user-input `id_name` might use the pattern

```rust
format!("^(?:id|ref)={};", id_name)
```

where `id_name` could include characters like `$`, `+`, or `?`. These characters would get interpreted as special regex sequences instead of plain text. Escaping the dynamic value would prevent this:

```rust
format!("^(?:id|ref)={};", fancy_regex::escape(id_name))
```

Currently I use the `fancy_regex` crate for matching and the `regex` crate for quoting, but this is error-prone, since `fancy_regex` adds a number of [regex-pattern syntax-extensions](https://docs.rs/fancy-regex/0.4.0/fancy_regex/index.html#syntax) which are not handled by `regex::escape`.

**Comment by robinst** [comment]
Yeah makes sense to have. @johnw42 recently worked on expansion APIs (a little bit related), maybe this is something they would be interested in adding.

> Currently I use the fancy_regex crate for matching and the regex crate for quoting, but this is error-prone, since fancy_regex adds a number of regex-pattern syntax-extensions which are not handled by regex::escape.

Which syntax causes problems in this? I'm asking because I'd expect regex crate's quoting to take care of them already, e.g. quoting all `\` and `(`.

**Comment by florianpircher** [comment]
> Which syntax causes problems in this? I'm asking because I'd expect regex crate's quoting to take care of them already, e.g. quoting all `\` and `(`.

The existing `regex::escape` function handles in my testing all of the current fancy extensions:

```rust
let s = r#"\h \H \e \1 \2 (?<name1>exp) \k<name1> (?P<name2>exp) (?P=name2) (?=exp) (?!exp) (?<=exp) (?<!exp) (?>exp)"#;
let e = regex::escape(s);
let re = fancy_regex::Regex::new(&e).unwrap();
assert_eq!(s, re.find(s).unwrap().unwrap().as_str());
```

The escaped value `e` looks like this:

```
\\h \\H \\e \\1 \\2 \(\?<name1>exp\) \\k<name1> \(\?P<name2>exp\) \(\?P=name2\) \(\?=exp\) \(\?!exp\) \(\?<=exp\) \(\?<!exp\) \(\?>exp\)
```

While this does work, it may not with future extensions to the `fancy-regex` crate (or to the `regex` crate). Using the escape function of one crate for another feels like something bound to break.

Maybe `fancy-regex` can provide a proxy `fancy_regex::escape` function which forwards the result from `regex::escape` as long as the escaped output stays compatible. If any of the two crates adds incompatible syntax-extensions the implementation of `fancy_regex::escape` can be changed to accommodate that.

**Comment by johnw42** [comment]
It's a bit unfortunate that the function I added for expansion templates is called `quote` rather than `escape`, but I don't know if it would be worth trying to change the name at this point.

The set of metacharacters is effectively part of the API of `regex` and `fancy-regex`, because changing it would change the meaning of existing expressions. Depending on `regex::escape` is therefore safe from a forward compatibility perspective.
It does feel messy, though, so IMHO it would be nice to make `fancy_regex::escape` work as well. Adding an `escape` function that either forwards to `regex::escape` or calls the private `push_quoted` function would be pretty trivial. I coded a version that returns `Cow<str>` like `Expander::quote` does; I'll put up a PR if anyone is interested.

**Comment by florianpircher** [comment]
> It's a bit unfortunate that the function I added for expansion templates is called `quote` rather than `escape`, but I don't know if it would be worth trying to change the name at this point.

If you want to change the name, you could still keep the old `quote` function around and hide it from the docs. This is what the `regex` crates does with [`captures_read_at` and `read_captures_at`](https://docs.rs/regex/1.4.1/src/regex/re_unicode.rs.html#677-707). I don’t feel strongly about this, keeping the `quote` name is OK, too.

> Adding an escape function that either forwards to regex::escape or calls the private push_quoted function would be pretty trivial. I coded a version that returns Cow<str> like Expander::quote does; I'll put up a PR if anyone is interested.

Sounds great, thank you!

**Comment by robinst** [comment]
> If you want to change the name, you could still keep the old quote function around and hide it from the docs. This is what the regex crates does with captures_read_at and read_captures_at. I don’t feel strongly about this, keeping the quote name is OK, too.

Yeah good idea. I don't mind the name `quote` tbh, so I don't feel strongly about this. I don't think it's an often-used function either (the one on expander).

> I'll put up a PR if anyone is interested.

Yes please!

**Comment by johnw42** [comment]
Here's the PR: https://github.com/fancy-regex/fancy-regex/pull/61

**Comment by robinst** [comment]
Released in version 0.4.1 🎉
